### ISSUE TITLE
Awkward error message

### ISSUE THREAD

**Issue by dselsam** [leanprover/lean3#669]
Hi,

The following silly code:

``` Lean
definition f {T : Prop} (t : T) := t
check f bool.tt
```

gives the following awkward error message:

```
error: type error in placeholder assigned to
  bool
placeholder has type
  Type₁
but is expected to have type
  Prop
the assignment was attempted when trying to solve
    type mismatch at application
      bool.tt
    term
      bool.tt
    has type
      bool
    but is expected to have type
      bool
```

The first half makes sense, but I suspect the second half is a (soft) error.

**Comment by leodemoura** [comment]
I get a slightly different message:

``` lean
/home/leo/projects/lean/build/release/bug5.lean:5:6: error: type error in placeholder assigned to
  bool
placeholder has type
  Type₁
but is expected to have type
  Prop
the assignment was attempted when trying to solve
    type mismatch at application
      f bool.tt
    term
      bool.tt
    has type
      bool
    but is expected to have type
      bool
```

Not that there is a `f` on my error message. Could you confirm whether you have the `f` or not?

BTW, if we use `set_option pp.all true`, then the message is slightly better, we get `@f bool bool.tt` instead of `f bool.tt`. I think it is less mysterious because the term `bool` appears in the second part.

The error is triggered when we try to assign the term `bool` to the metavariable `?M` in `@f ?M bool.tt`.
The cryptic error message is trying to explain why the placeholder was assigned to `bool`.

**Comment by dselsam** [comment]
> Could you confirm whether you have the f or not?

I confirm that I do have `f`. Sorry -- the error message I posted must have been from 

``` Lean
check (λ {T : Prop} (t : T), t) bool.tt
```

(λs don't seem to print in that slot unless `pp.all` is set to `true` ).

> The error is triggered when we try to assign the term bool to the metavariable ?M in @f ?M bool.tt. The cryptic error message is trying to explain why the placeholder was assigned to bool.

I understand the user error. Only the last part of the error message seems weird:

```
term
      bool.tt
    has type
      bool
    but is expected to have type
      bool
```

This makes it seem like it tried to unify the type of `bool.tt` with `bool` and failed, which is not what happened. I think the error message up until that point explains the problem adequately though.

**Comment by leodemoura** [comment]
> I confirm that I do have f. Sorry -- the error message I posted must have been from
> 
> check (λ {T : Prop} (t : T), t) bool.tt
> (λs don't seem to print in that slot unless pp.all is set to true ).

I see. This is indeed really weird. The problem is that the pretty printer applies beta-reduction before printing a term. 
This feature is useful when displaying proof goals, but it seems to be a bad idea for error messages.
I will disable it for error messages.

> Only the last part of the error message seems weird:

I will think about it and how to improve it.

**Comment by leodemoura** [comment]
The commits above improved the error message.
Given

``` lean
check (λ {T : Prop} (t : T), t) bool.tt
```

We get the following error

```
669.lean:1:9: error: type error in placeholder assigned to
  bool
placeholder has type
  Type₁
but is expected to have type
  Prop
the assignment was attempted when processing
    application type constraint
      (λ {T : Prop} (t : T), t) bool.tt
    term
      bool.tt
    has type
      bool
```

**Comment by dselsam** [comment]
I think this is much clearer. Thanks.

**Comment by dselsam** [comment]
Hi,

I think the fix in 7bad9fe might not always be the right thing to do. The reason the beta-reduction was confusing in the original example was that the application that was reduced was not type correct, and so the cause of the error was hidden by the beta-reduction. 

However, consider the following error message that I just got:

``` Lean
term
  unique
has type
  ∀ (Y1 Y2 : set T),
    Y1 ∈ X ∧ u ∈ Y1 → Y2 ∈ X ∧ u ∈ Y2 → Y2 = Y1
but is expected to have type
  ∀ (u_1 v : set T),
    (λ (Y : set T), Y ∈ X ∧ u ∈ Y) u_1 → (λ (Y : set T), Y ∈ X ∧ u ∈ Y) v → u_1 = v
```

The lambdas obfuscate the error, and the random binder names compound the issue. It took me a minute to figure out that I had just switched the order in the final equality.

I think the error taxonomy is actually quite rich and subtle, and it may not be worth it to play whack-a-mole right now. Over the next few months I'll do a more systematic study of the possible errors, and will try to make a detailed proposal.

**Comment by leodemoura** [comment]
> I think the error taxonomy is actually quite rich and subtle, and it may not be worth it to play whack-a-mole right now. Over the next few months I'll do a more systematic study of the possible errors, and will try to make a detailed proposal.

Thanks.
